### PR TITLE
Added ability to captchalouge items inside any inventory screen.

### PR DESCRIPTION
**Pull request by seriousCoolness**
To captchalouge something from inside your inventory, hover over it and press the captchalouge button.

### CODE REVIEW

File: src/com/mraof/minestuck/client/settings/MinestuckKeyHandler.java [hunk 1]
```diff
@@ -1,11 +1,16 @@
...
 import net.minecraft.client.Minecraft;
+import net.minecraft.client.gui.inventory.GuiContainer;
+import net.minecraft.client.gui.inventory.GuiInventory;
 import net.minecraft.client.settings.KeyBinding;
 import net.minecraftforge.fml.client.registry.ClientRegistry;
 import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
 import net.minecraftforge.fml.common.gameevent.TickEvent;
 
+import org.lwjgl.input.Keyboard;
+import org.lwjgl.input.Mouse;
```
Comment (kirderf1) [inline src/com/mraof/minestuck/client/settings/MinestuckKeyHandler.java]: Are these imports needed?

File: src/com/mraof/minestuck/client/settings/MinestuckKeyHandler.java [hunk 2]
```diff
@@ -52,9 +58,15 @@ else if(captchaKey.isPressed())
...
 		}
+		else if(Keyboard.isKeyDown(captchaKey.getKeyCode()) && !captchaKeyPressed) {
+			if(Minecraft.getMinecraft().currentScreen instanceof GuiContainer && ((GuiContainer)Minecraft.getMinecraft().currentScreen).getSlotUnderMouse().getHasStack())
+			MinestuckChannelHandler.sendToServer(MinestuckPacket.makePacket(MinestuckPacket.Type.CAPTCHA, CaptchaDeckPacket.CAPTCHALOUGE_INV));
+		}
+		
 		
 		statKeyPressed = statKey.isKeyDown();
 		editKeyPressed = editKey.isKeyDown();
+		captchaKeyPressed = Keyboard.isKeyDown(captchaKey.getKeyCode());
```
Comment (kirderf1) [inline src/com/mraof/minestuck/client/settings/MinestuckKeyHandler.java]: Why is this not using the same function as the lines above?

File: src/com/mraof/minestuck/inventory/captchalouge/CaptchaDeckHandler.java [hunk 3]
```diff
@@ -12,9 +12,16 @@
 import com.mraof.minestuck.alchemy.AlchemyRecipes;
 import com.mraof.minestuck.util.Debug;
 import com.mraof.minestuck.util.MinestuckPlayerData;
+
+import net.minecraft.client.Minecraft;
+import net.minecraft.client.gui.GuiScreen;
+import net.minecraft.client.gui.inventory.GuiContainer;
+import net.minecraft.client.gui.inventory.GuiInventory;
```
Comment (kirderf1) [inline src/com/mraof/minestuck/inventory/captchalouge/CaptchaDeckHandler.java]: I don't think that this class is intended to involve itself with guis

File: src/com/mraof/minestuck/inventory/captchalouge/CaptchaDeckHandler.java [hunk 4]
```diff
@@ -254,6 +261,66 @@ else if(card1 && card2)
 		
 	}
 	
+	public static void captchalougeInventoryItem (EntityPlayerMP player) {
+		GuiScreen guiScreen = Minecraft.getMinecraft().currentScreen;
```
Comment (kirderf1) [inline src/com/mraof/minestuck/inventory/captchalouge/CaptchaDeckHandler.java]: This approach will not work as guis aren't available on the server side, while this function clearly is on the server side

File: src/com/mraof/minestuck/network/CaptchaDeckPacket.java [hunk 5]
```diff
@@ -120,6 +124,8 @@ public void execute(EntityPlayer player)
 				CaptchaDeckHandler.useItem((EntityPlayerMP) player);
 			else if(this.type == CAPTCHALOUGE && !player.getItemStackFromSlot(EntityEquipmentSlot.MAINHAND).isEmpty())
 				CaptchaDeckHandler.captchalougeItem((EntityPlayerMP) player);
+			else if(this.type == CAPTCHALOUGE_INV)
+				CaptchaDeckHandler.captchalougeInventoryItem((EntityPlayerMP) player);
```
Comment (kirderf1) [inline src/com/mraof/minestuck/network/CaptchaDeckPacket.java]: Unless the server side keeps track of which item that is mouseovered, you need to send that information with this packet.

**Comment by kirderf1** [comment]
On another note, an alternative to captchaloguing the item mouseovered, there's the option of captchaloguing the item held by the mouse. Has this alternative been considered and compared to the current idea?

**Comment by kirderf1** [comment]
Testing the code out, I was unable to captchalogue from the toolbar, but it appeared to work with the rest of the inventory. Not the rest of the gui though, since when I accidentaly pressed c without hovering over a slot, the game crashed.
`java.lang.NullPointerException: Unexpected error
	at com.mraof.minestuck.client.settings.MinestuckKeyHandler.onTick(MinestuckKeyHandler.java:62)
	at net.minecraftforge.fml.common.eventhandler.ASMEventHandler_17_MinestuckKeyHandler_onTick_ClientTickEvent.invoke(.dynamic)
...`
It's not ideal to add new things to a pull request that wasn't originally supposed to be in the pull request, in this case that new weapon. Not great, but you'll just have to keep that in mind for next time.
Also, I would still like to have an answer on the previous comment.

**Comment by seriousCoolness** [comment]
> 
> 
> Testing the code out, I was unable to captchalogue from the toolbar, but it appeared to work with the rest of the inventory. Not the rest of the gui though, since when I accidentaly pressed c without hovering over a slot, the game crashed.
> `java.lang.NullPointerException: Unexpected error at com.mraof.minestuck.client.settings.MinestuckKeyHandler.onTick(MinestuckKeyHandler.java:62) at net.minecraftforge.fml.common.eventhandler.ASMEventHandler_17_MinestuckKeyHandler_onTick_ClientTickEvent.invoke(.dynamic) ...`
> It's not ideal to add new things to a pull request that wasn't originally supposed to be in the pull request, in this case that new weapon. Not great, but you'll just have to keep that in mind for next time.
> Also, I would still like to have an answer on the previous comment.

Sorry about the new weapon! I didn't realize how the pull requests worked and I thought i could somehow set the request to ignore certain commits, but when i found out I couldn't I must have forgotten to notify you of it. I'm still pretty inexperienced with Git, so apologies for any mistakes I make.
To answer your other question, I considered it, but I think that having to click the item and then press C would counteract the convenience of the shortcut somewhat. Although considering how easy it would be, I could probably just add it along with the final fixes to the current method. That said, feel free to offer your opinion and feelings on the matter, as convenience is extremely subjective, obviously.

(Also oh my god I cant believe i somehow deleted the _three different statements_ that were _specifically written to make sure nullref errors wouldn't happen._ I really need to learn to not be lazy and impatient with my testing.)

**Comment by kirderf1** [comment]
Alright.
Anyways, toolbar captchaloguing still doesn't work entierly. It does only let you captchalogue when mouseovering an item, but it appears to captchalogue the item to the left.
(Hope you can test it out and make sure it works next time! Would save a bit of time for going back and forth)

**Pull request merged**